pkg/common/cns-lib/volume/manager.go

```diff
@@ -318,23 +318,67 @@ func (m *defaultManager) createVolumeWithImprovedIdempotency(ctx context.Context
 		}
 	}
 
-	// Get the taskInfo.
-	taskInfo, err := cns.GetTaskInfo(ctx, task)
-	if err != nil || taskInfo == nil {
-		log.Errorf("failed to get taskInfo for CreateVolume task with err: %v", err)
-		if err != nil {
-			faultType = ExtractFaultTypeFromErr(ctx, err)
-		} else {
-			faultType = csifault.CSITaskInfoEmptyFault
+	taskInfo, err := task.WaitForResult(ctx, nil)
+	if err != nil {
+		if cnsvsphere.IsManagedObjectNotFound(err, task.Reference()) {
+			log.Debugf("CreateVolume task %s not found in VCDB. Querying CNS "+
+				"to determine if the volume %s was successfully created.",
+				task.Reference().Value, volNameFromInputSpec)
+			queryFilter := cnstypes.CnsQueryFilter{
+				Names:               []string{volNameFromInputSpec},
+				ContainerClusterIds: []string{spec.Metadata.ContainerClusterArray[0].ClusterId},
+			}
+			querySelection := cnstypes.CnsQuerySelection{
+				Names: []string{
+					string(cnstypes.QuerySelectionNameTypeVolumeName),
+				},
+			}
+			queryResult, queryAllVolumeErr := m.QueryAllVolume(ctx, queryFilter, querySelection)
+			if queryAllVolumeErr != nil {
+				log.Debugf("failed to query CNS for volume %s with error: %v. Cannot "+
+					"determine if CreateVolume task %s was successful.", volNameFromInputSpec,
+					queryAllVolumeErr, task.Reference().Value)
+				return nil, ExtractFaultTypeFromErr(ctx, err), err
+			}
+			if len(queryResult.Volumes) != 0 {
+				volumeID := queryResult.Volumes[0].VolumeId.Id
+				volumeOperationDetails = createRequestDetails(volNameFromInputSpec, volumeID, "", 0,
+					volumeOperationDetails.OperationDetails.TaskInvocationTimestamp, task.Reference().Value,
+					volumeOperationDetails.OperationDetails.TaskID, taskInvocationStatusSuccess, "")
+				log.Debugf("Found volume %s with id %s", volNameFromInputSpec, volumeID)
+				return &CnsVolumeInfo{
+					DatastoreURL: "",
+					VolumeID: cnstypes.CnsVolumeId{
+						Id: volumeID,
+					},
+				}, "", nil
+			}
+			log.Errorf("volume with name %s not present in CNS. Marking task %s as failed.",
+				volNameFromInputSpec, task.Reference().Value)
+			volumeOperationDetails = createRequestDetails(volNameFromInputSpec, "", "", 0,
+				volumeOperationDetails.OperationDetails.TaskInvocationTimestamp, task.Reference().Value,
+				volumeOperationDetails.OperationDetails.TaskID, taskInvocationStatusError, err.Error())
+
+			return nil, ExtractFaultTypeFromErr(ctx, err), err
 		}
-		return nil, faultType, err
+		// WaitForResult can fail for many reasons, including:
+		// - CNS restarted and marked "InProgress" tasks as "Failed".
+		// - Any failures from CNS.
+		// - Any failures at lower layers like FCD and SPS.
+		// In all cases, mark task as failed and retry.
```

Comment:
Does this "retry" refer to retries by the external-provisioner or will CSI driver attempt to retry?

Reply:
Retry refers to external-provisioner. CSI will return an error to the provisioner